Use Navicat client open database, console print ClassCastException error log.

## Bug Report

**For English only**, other languages will not accept.

Before report a bug, make sure you have:

- Searched open and closed [GitHub issues](https://github.com/apache/incubator-shardingsphere/issues).
- Read documentation: [ShardingSphere Doc](https://shardingsphere.apache.org/document/current/en/overview).

Please pay attention on issues you submitted, because we maybe need more details. 
If no response **more than 7 days** and we cannot reproduce it on current information, we will **close it**.

Please answer these questions before submitting your issue. Thanks!

### Which version of ShardingSphere did you use?
Lastest master code
### Which project did you use? Sharding-JDBC or Sharding-Proxy?
Sharding-Proxy
### Expected behavior
Use Navicat client, open database.
### Actual behavior
Can open database, but console print error log.
### Reason analyze (If you can)

### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.

### Example codes for reproduce this issue (such as a github link).

```
[INFO ] 12:23:05.330 [nioEventLoopGroup-2-1] i.n.handler.logging.LoggingHandler - [id: 0x1b9ce6ac, L:/0:0:0:0:0:0:0:0:3308] READ: [id: 0x2d2ad5dd, L:/127.0.0.1:3308 - R:/127.0.0.1:50754]
[INFO ] 12:23:05.331 [nioEventLoopGroup-2-1] i.n.handler.logging.LoggingHandler - [id: 0x1b9ce6ac, L:/0:0:0:0:0:0:0:0:3308] READ COMPLETE
[INFO ] 12:23:06.075 [ShardingSphere-Command-0] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:06.075 [ShardingSphere-Command-0] ShardingSphere-SQL - Logic SQL: SET NAMES utf8mb4
[INFO ] 12:23:06.075 [ShardingSphere-Command-0] ShardingSphere-SQL - SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.SetStatement@3e5a50af, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@2c15d279)
[INFO ] 12:23:06.076 [ShardingSphere-Command-0] ShardingSphere-SQL - Actual SQL: ds_0 ::: SET NAMES utf8mb4
[INFO ] 12:23:06.076 [ShardingSphere-Command-0] ShardingSphere-SQL - Actual SQL: ds_1 ::: SET NAMES utf8mb4
[INFO ] 12:23:06.164 [ShardingSphere-Command-1] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:06.164 [ShardingSphere-Command-1] ShardingSphere-SQL - Logic SQL: SHOW VARIABLES LIKE 'lower_case_%'
[INFO ] 12:23:06.164 [ShardingSphere-Command-1] ShardingSphere-SQL - SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.dialect.mysql.ShowOtherStatement@46335750, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@8b98b9b)
[INFO ] 12:23:06.164 [ShardingSphere-Command-1] ShardingSphere-SQL - Actual SQL: ds_1 ::: SHOW VARIABLES LIKE 'lower_case_%'
[INFO ] 12:23:10.510 [nioEventLoopGroup-2-1] i.n.handler.logging.LoggingHandler - [id: 0x1b9ce6ac, L:/0:0:0:0:0:0:0:0:3308] READ: [id: 0xca3fe6d8, L:/127.0.0.1:3308 - R:/127.0.0.1:50807]
[INFO ] 12:23:10.511 [nioEventLoopGroup-2-1] i.n.handler.logging.LoggingHandler - [id: 0x1b9ce6ac, L:/0:0:0:0:0:0:0:0:3308] READ COMPLETE
[INFO ] 12:23:10.521 [ShardingSphere-Command-5] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:10.521 [ShardingSphere-Command-5] ShardingSphere-SQL - Logic SQL: SET NAMES utf8mb4
[INFO ] 12:23:10.521 [ShardingSphere-Command-5] ShardingSphere-SQL - SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.SetStatement@5542c1ce, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@75a4a8d3)
[INFO ] 12:23:10.521 [ShardingSphere-Command-5] ShardingSphere-SQL - Actual SQL: ds_0 ::: SET NAMES utf8mb4
[INFO ] 12:23:10.521 [ShardingSphere-Command-5] ShardingSphere-SQL - Actual SQL: ds_1 ::: SET NAMES utf8mb4
[INFO ] 12:23:10.575 [ShardingSphere-Command-7] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:10.576 [ShardingSphere-Command-7] ShardingSphere-SQL - Logic SQL: SHOW TABLE STATUS
[INFO ] 12:23:10.576 [ShardingSphere-Command-7] ShardingSphere-SQL - SQLStatement: CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dal.dialect.mysql.ShowTableStatusStatement@37669c66, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@33590e52)
[INFO ] 12:23:10.576 [ShardingSphere-Command-7] ShardingSphere-SQL - Actual SQL: ds_1 ::: SHOW TABLE STATUS
[INFO ] 12:23:10.616 [ShardingSphere-Command-4] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:10.617 [ShardingSphere-Command-4] ShardingSphere-SQL - Logic SQL: SELECT COUNT(*) FROM information_schema.TABLES WHERE TABLE_SCHEMA = 'databasemanager' UNION SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = 'databasemanager' UNION SELECT COUNT(*) FROM information_schema.ROUTINES WHERE ROUTINE_SCHEMA = 'databasemanager'
[INFO ] 12:23:10.617 [ShardingSphere-Command-4] ShardingSphere-SQL - SQLStatement: SelectStatementContext(super=CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dml.SelectStatement@3f763a5b, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@7ae644d8), tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@7ae644d8, projectionsContext=ProjectionsContext(startIndex=7, stopIndex=14, distinctRow=false, projections=[AggregationProjection(type=COUNT, innerExpression=(*), alias=Optional.empty, derivedAggregationProjections=[], index=-1)], columnLabels=[COUNT(*)]), groupByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.groupby.GroupByContext@7ee6828c, orderByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.orderby.OrderByContext@532eca15, paginationContext=org.apache.shardingsphere.sql.parser.relation.segment.select.pagination.PaginationContext@5d8496c9, containsSubquery=false)
[INFO ] 12:23:10.617 [ShardingSphere-Command-4] ShardingSphere-SQL - Actual SQL: ds_1 ::: SELECT COUNT(*) FROM information_schema.TABLES WHERE TABLE_SCHEMA = 'databasemanager' UNION SELECT COUNT(*) FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = 'databasemanager' UNION SELECT COUNT(*) FROM information_schema.ROUTINES WHERE ROUTINE_SCHEMA = 'databasemanager'
[INFO ] 12:23:10.671 [ShardingSphere-Command-8] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:10.671 [ShardingSphere-Command-8] ShardingSphere-SQL - Logic SQL: SELECT TABLE_SCHEMA, TABLE_NAME, TABLE_TYPE FROM information_schema.TABLES WHERE TABLE_SCHEMA = 'databasemanager' ORDER BY TABLE_SCHEMA, TABLE_TYPE
[INFO ] 12:23:10.671 [ShardingSphere-Command-8] ShardingSphere-SQL - SQLStatement: SelectStatementContext(super=CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dml.SelectStatement@5139217f, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@4386d13), tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@4386d13, projectionsContext=ProjectionsContext(startIndex=7, stopIndex=42, distinctRow=false, projections=[ColumnProjection(owner=null, name=TABLE_SCHEMA, alias=Optional.empty), ColumnProjection(owner=null, name=TABLE_NAME, alias=Optional.empty), ColumnProjection(owner=null, name=TABLE_TYPE, alias=Optional.empty)], columnLabels=[TABLE_SCHEMA, TABLE_NAME, TABLE_TYPE]), groupByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.groupby.GroupByContext@3db4633e, orderByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.orderby.OrderByContext@7c134aa6, paginationContext=org.apache.shardingsphere.sql.parser.relation.segment.select.pagination.PaginationContext@6439ae82, containsSubquery=false)
[INFO ] 12:23:10.671 [ShardingSphere-Command-8] ShardingSphere-SQL - Actual SQL: ds_1 ::: SELECT TABLE_SCHEMA, TABLE_NAME, TABLE_TYPE FROM information_schema.TABLES WHERE TABLE_SCHEMA = 'databasemanager' ORDER BY TABLE_SCHEMA, TABLE_TYPE
[INFO ] 12:23:10.690 [ShardingSphere-Command-9] ShardingSphere-SQL - Rule Type: sharding
[INFO ] 12:23:10.690 [ShardingSphere-Command-9] ShardingSphere-SQL - Logic SQL: SELECT TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME, COLUMN_TYPE FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = 'databasemanager' ORDER BY TABLE_SCHEMA, TABLE_NAME
[INFO ] 12:23:10.690 [ShardingSphere-Command-9] ShardingSphere-SQL - SQLStatement: SelectStatementContext(super=CommonSQLStatementContext(sqlStatement=org.apache.shardingsphere.sql.parser.sql.statement.dml.SelectStatement@4a6ca2d2, tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@f6e35fb), tablesContext=org.apache.shardingsphere.sql.parser.relation.segment.table.TablesContext@f6e35fb, projectionsContext=ProjectionsContext(startIndex=7, stopIndex=56, distinctRow=false, projections=[ColumnProjection(owner=null, name=TABLE_SCHEMA, alias=Optional.empty), ColumnProjection(owner=null, name=TABLE_NAME, alias=Optional.empty), ColumnProjection(owner=null, name=COLUMN_NAME, alias=Optional.empty), ColumnProjection(owner=null, name=COLUMN_TYPE, alias=Optional.empty)], columnLabels=[TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME, COLUMN_TYPE]), groupByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.groupby.GroupByContext@6951246d, orderByContext=org.apache.shardingsphere.sql.parser.relation.segment.select.orderby.OrderByContext@3eabbbb8, paginationContext=org.apache.shardingsphere.sql.parser.relation.segment.select.pagination.PaginationContext@117ef48d, containsSubquery=false)
[INFO ] 12:23:10.691 [ShardingSphere-Command-9] ShardingSphere-SQL - Actual SQL: ds_1 ::: SELECT TABLE_SCHEMA, TABLE_NAME, COLUMN_NAME, COLUMN_TYPE FROM information_schema.COLUMNS WHERE TABLE_SCHEMA = 'databasemanager' ORDER BY TABLE_SCHEMA, TABLE_NAME
[ERROR] 12:23:10.916 [ShardingSphere-Command-10] o.a.s.s.f.c.CommandExecutorTask - Exception occur: 
java.lang.ClassCastException: org.apache.shardingsphere.sql.parser.sql.segment.dml.expr.subquery.SubquerySegment cannot be cast to org.apache.shardingsphere.sql.parser.sql.segment.generic.TableSegment
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitJoinedTable(MySQLDMLVisitor.java:515)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitJoinedTable(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$JoinedTableContext.accept(MySQLStatementParser.java:6736)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.getTableSegments(MySQLDMLVisitor.java:338)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitTableReference(MySQLDMLVisitor.java:490)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitTableReference(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$TableReferenceContext.accept(MySQLStatementParser.java:6265)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitEscapedTableReference(MySQLDMLVisitor.java:476)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitEscapedTableReference(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$EscapedTableReferenceContext.accept(MySQLStatementParser.java:5936)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitTableReferences(MySQLDMLVisitor.java:469)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitTableReferences(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$TableReferencesContext.accept(MySQLStatementParser.java:5880)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitFromClause(MySQLDMLVisitor.java:461)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitFromClause(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$FromClauseContext.accept(MySQLStatementParser.java:5835)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitSelectClause(MySQLDMLVisitor.java:315)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitSelectClause(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$SelectClauseContext.accept(MySQLStatementParser.java:4542)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitUnionClause(MySQLDMLVisitor.java:303)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitUnionClause(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$UnionClauseContext.accept(MySQLStatementParser.java:4437)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitSelect(MySQLDMLVisitor.java:295)
	at org.apache.shardingsphere.sql.parser.visitor.impl.MySQLDMLVisitor.visitSelect(MySQLDMLVisitor.java:119)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$SelectContext.accept(MySQLStatementParser.java:3095)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.SQLParserEngine.parse0(SQLParserEngine.java:72)
	at org.apache.shardingsphere.sql.parser.SQLParserEngine.parse(SQLParserEngine.java:53)
	at org.apache.shardingsphere.shardingproxy.backend.text.TextProtocolBackendHandlerFactory.newInstance(TextProtocolBackendHandlerFactory.java:69)
	at org.apache.shardingsphere.shardingproxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.<init>(MySQLComQueryPacketExecutor.java:65)
	at org.apache.shardingsphere.shardingproxy.frontend.mysql.command.MySQLCommandExecutorFactory.newInstance(MySQLCommandExecutorFactory.java:70)
	at org.apache.shardingsphere.shardingproxy.frontend.mysql.command.MySQLCommandExecuteEngine.getCommandExecutor(MySQLCommandExecuteEngine.java:58)
	at org.apache.shardingsphere.shardingproxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:89)
	at org.apache.shardingsphere.shardingproxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:70)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

--- a/encrypt-core/encrypt-core-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/impl/EncryptProjectionTokenGenerator.java
+++ b/encrypt-core/encrypt-core-rewrite/src/main/java/org/apache/shardingsphere/encrypt/rewrite/token/generator/impl/EncryptProjectionTokenGenerator.java
@@ -58,7 +58,7 @@ public final class EncryptProjectionTokenGenerator extends BaseEncryptSQLTokenGe
     public Collection<SubstitutableColumnNameToken> generateSQLTokens(final SelectStatementContext selectStatementContext) {
         ProjectionsSegment projectionsSegment = selectStatementContext.getSqlStatement().getProjections();
         // TODO process multiple tables
-        String tableName = selectStatementContext.getSqlStatement().getTables().iterator().next().getTableName().getIdentifier().getValue();
+        String tableName = selectStatementContext.getSqlStatement().getSimpleTableSegments().iterator().next().getTableName().getIdentifier().getValue();
         return getEncryptRule().findEncryptTable(tableName).map(
             encryptTable -> generateSQLTokens(projectionsSegment, tableName, selectStatementContext, encryptTable)).orElseGet(Collections::emptyList);
     }

--- a/sharding-core/sharding-core-rewrite/src/main/java/org/apache/shardingsphere/sharding/rewrite/token/generator/impl/OrderByTokenGenerator.java
+++ b/sharding-core/sharding-core-rewrite/src/main/java/org/apache/shardingsphere/sharding/rewrite/token/generator/impl/OrderByTokenGenerator.java
@@ -67,7 +67,7 @@ public final class OrderByTokenGenerator implements OptionalSQLTokenGenerator<Se
         if (selectStatement.getWhere().isPresent()) {
             return selectStatement.getWhere().get().getStopIndex() + 1;
         } else {
-            return selectStatement.getTables().stream().mapToInt(SimpleTableSegment::getStopIndex).max().getAsInt() + 1;
+            return selectStatement.getSimpleTableSegments().stream().mapToInt(SimpleTableSegment::getStopIndex).max().getAsInt() + 1;
         }
     }
 }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/sql/statement/dml/SelectStatement.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-engine/src/main/java/org/apache/shardingsphere/sql/parser/sql/statement/dml/SelectStatement.java
@@ -26,6 +26,8 @@ import org.apache.shardingsphere.sql.parser.sql.segment.dml.pagination.limit.Lim
 import org.apache.shardingsphere.sql.parser.sql.segment.dml.predicate.LockSegment;
 import org.apache.shardingsphere.sql.parser.sql.segment.dml.predicate.WhereSegment;
 import org.apache.shardingsphere.sql.parser.sql.segment.generic.table.SimpleTableSegment;
+import org.apache.shardingsphere.sql.parser.sql.segment.generic.table.SubqueryTableSegment;
+import org.apache.shardingsphere.sql.parser.sql.segment.generic.table.TableSegment;
 
 import java.util.Collection;
 import java.util.LinkedList;
@@ -38,7 +40,7 @@ import java.util.Optional;
 @Setter
 public final class SelectStatement extends DMLStatement {
     
-    private final Collection<SimpleTableSegment> tables = new LinkedList<>();
+    private final Collection<TableSegment> tables = new LinkedList<>();
     
     private ProjectionsSegment projections;
     
@@ -97,5 +99,35 @@ public final class SelectStatement extends DMLStatement {
      */
     public Optional<LockSegment> getLock() {
         return Optional.ofNullable(lock);
+    }
+    
+    /**
+     * Get simple table segments.
+     * 
+     * @return simple table segments
+     */
+    public Collection<SimpleTableSegment> getSimpleTableSegments() {
+        Collection<SimpleTableSegment> result = new LinkedList<>();
+        for (TableSegment each : tables) {
+            if (each instanceof SimpleTableSegment) {
+                result.add((SimpleTableSegment) each);
+            }
+        }
+        return result;
+    }
+    
+    /**
+     * Get subQuery table segments.
+     *
+     * @return subQuery table segments
+     */
+    public Collection<SubqueryTableSegment> getSubQueryTableSegments() {
+        Collection<SubqueryTableSegment> result = new LinkedList<>();
+        for (TableSegment each : tables) {
+            if (each instanceof SubqueryTableSegment) {
+                result.add((SubqueryTableSegment) each);
+            }
+        }
+        return result;
     }
 }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/MySQLDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/MySQLDMLVisitor.java
@@ -481,6 +481,7 @@ public final class MySQLDMLVisitor extends MySQLVisitor implements DMLVisitor {
     public ASTNode visitTableReference(final TableReferenceContext ctx) {
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         if (null != ctx.tableFactor()) {
+            // TODO :Ignore subquery table segment
             ASTNode tableFactor = visit(ctx.tableFactor());
             if (tableFactor instanceof SimpleTableSegment) {
                 result.getValue().add((SimpleTableSegment) tableFactor);
@@ -516,13 +517,17 @@ public final class MySQLDMLVisitor extends MySQLVisitor implements DMLVisitor {
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
+        // TODO: Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
-        SimpleTableSegment tableSegment = (SimpleTableSegment) visit(ctx.tableFactor());
-        result.getValue().add(tableSegment);
+        ASTNode tableFactor = visit(ctx.tableFactor());
+        if (tableFactor instanceof SubqueryTableSegment) {
+            return result;
+        }
+        result.getValue().add((SimpleTableSegment) tableFactor);
         if (null != ctx.joinSpecification()) {
             Collection<SimpleTableSegment> tableSegments = new LinkedList<>();
             for (SimpleTableSegment each : ((CollectionValue<SimpleTableSegment>) visit(ctx.joinSpecification())).getValue()) {
-                if (isTable(each, Collections.singleton(tableSegment))) {
+                if (isTable(each, Collections.singleton((SimpleTableSegment) tableFactor))) {
                     tableSegments.add(each);
                 }
             }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/MySQLDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/MySQLDMLVisitor.java
@@ -397,7 +397,7 @@ public final class MySQLDMLVisitor extends MySQLVisitor implements DMLVisitor {
     
     @Override
     public ASTNode visitProjection(final ProjectionContext ctx) {
-        // FIXME: The stop index of project is the stop index of projection, instead of alias.
+        // FIXME :The stop index of project is the stop index of projection, instead of alias.
         if (null != ctx.qualifiedShorthand()) {
             QualifiedShorthandContext shorthand = ctx.qualifiedShorthand();
             ShorthandProjectionSegment result = new ShorthandProjectionSegment(shorthand.getStart().getStartIndex(), shorthand.getStop().getStopIndex(), shorthand.getText());

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/MySQLDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-mysql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/MySQLDMLVisitor.java
@@ -293,7 +293,7 @@ public final class MySQLDMLVisitor extends MySQLVisitor implements DMLVisitor {
     
     @Override
     public ASTNode visitSelect(final SelectContext ctx) {
-        // TODO : Unsupported for withClause.
+        // TODO :Unsupported for withClause.
         SelectStatement result = (SelectStatement) visit(ctx.unionClause());
         result.setParameterCount(getCurrentParameterIndex());
         return result;
@@ -301,7 +301,7 @@ public final class MySQLDMLVisitor extends MySQLVisitor implements DMLVisitor {
     
     @Override
     public ASTNode visitUnionClause(final UnionClauseContext ctx) {
-        // TODO : Unsupported for union SQL.
+        // TODO :Unsupported for union SQL.
         return visit(ctx.selectClause(0));
     }
     
@@ -517,7 +517,7 @@ public final class MySQLDMLVisitor extends MySQLVisitor implements DMLVisitor {
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
-        // TODO: Bad processing for join table
+        // TODO :Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         ASTNode tableFactor = visit(ctx.tableFactor());
         if (tableFactor instanceof SubqueryTableSegment) {

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/OracleDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/OracleDMLVisitor.java
@@ -384,6 +384,7 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     public ASTNode visitTableReference(final TableReferenceContext ctx) {
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         if (null != ctx.tableFactor()) {
+            // TODO :Ignore subquery table segment
             ASTNode tableFactor = visit(ctx.tableFactor());
             if (tableFactor instanceof SimpleTableSegment) {
                 result.getValue().add((SimpleTableSegment) tableFactor);
@@ -419,13 +420,17 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
+        // TODO: Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
-        SimpleTableSegment tableSegment = (SimpleTableSegment) visit(ctx.tableFactor());
-        result.getValue().add(tableSegment);
+        ASTNode tableFactor = visit(ctx.tableFactor());
+        if (tableFactor instanceof SubqueryTableSegment) {
+            return result;
+        }
+        result.getValue().add((SimpleTableSegment) tableFactor);
         if (null != ctx.joinSpecification()) {
             Collection<SimpleTableSegment> tableSegments = new LinkedList<>();
             for (SimpleTableSegment each : ((CollectionValue<SimpleTableSegment>) visit(ctx.joinSpecification())).getValue()) {
-                if (isTable(each, Collections.singleton(tableSegment))) {
+                if (isTable(each, Collections.singleton((SimpleTableSegment) tableFactor))) {
                     tableSegments.add(each);
                 }
             }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/OracleDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/OracleDMLVisitor.java
@@ -305,7 +305,7 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     
     @Override
     public ASTNode visitProjection(final ProjectionContext ctx) {
-        // FIXME: The stop index of project is the stop index of projection, instead of alias.
+        // FIXME :The stop index of project is the stop index of projection, instead of alias.
         if (null != ctx.qualifiedShorthand()) {
             QualifiedShorthandContext shorthand = ctx.qualifiedShorthand();
             ShorthandProjectionSegment result = new ShorthandProjectionSegment(shorthand.getStart().getStartIndex(), shorthand.getStop().getStopIndex(), shorthand.getText());

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/OracleDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-oracle/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/OracleDMLVisitor.java
@@ -99,7 +99,7 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     @Override
     public ASTNode visitInsert(final InsertContext ctx) {
         // TODO :FIXME, since there is no segment for insertValuesClause, InsertStatement is created by sub rule.
-        // TODO deal with insert select
+        // TODO :deal with insert select
         if (null != ctx.insertSingleTable()) {
             InsertStatement result = (InsertStatement) visit(ctx.insertSingleTable().insertValuesClause());
             result.setTable((SimpleTableSegment) visit(ctx.insertSingleTable().insertIntoClause().tableName()));
@@ -224,7 +224,7 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     
     @Override
     public ASTNode visitSelect(final SelectContext ctx) {
-        // TODO : Unsupported for withClause.
+        // TODO :Unsupported for withClause.
         SelectStatement result = (SelectStatement) visit(ctx.unionClause());
         result.setParameterCount(getCurrentParameterIndex());
         return result;
@@ -232,7 +232,7 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     
     @Override
     public ASTNode visitUnionClause(final UnionClauseContext ctx) {
-        // TODO : Unsupported for union SQL.
+        // TODO :Unsupported for union SQL.
         return visit(ctx.selectClause(0));
     }
     
@@ -420,7 +420,7 @@ public final class OracleDMLVisitor extends OracleVisitor implements DMLVisitor 
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
-        // TODO: Bad processing for join table
+        // TODO :Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         ASTNode tableFactor = visit(ctx.tableFactor());
         if (tableFactor instanceof SubqueryTableSegment) {

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/PostgreSQLDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/PostgreSQLDMLVisitor.java
@@ -314,7 +314,7 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     
     @Override
     public ASTNode visitProjection(final ProjectionContext ctx) {
-        // FIXME: The stop index of project is the stop index of projection, instead of alias.
+        // FIXME :The stop index of project is the stop index of projection, instead of alias.
         if (null != ctx.qualifiedShorthand()) {
             QualifiedShorthandContext shorthand = ctx.qualifiedShorthand();
             ShorthandProjectionSegment result = new ShorthandProjectionSegment(shorthand.getStart().getStartIndex(), shorthand.getStop().getStopIndex(), shorthand.getText());

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/PostgreSQLDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/PostgreSQLDMLVisitor.java
@@ -108,7 +108,7 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     @Override
     public ASTNode visitInsert(final InsertContext ctx) {
         // TODO :FIXME, since there is no segment for insertValuesClause, InsertStatement is created by sub rule.
-        // TODO deal with insert select
+        // TODO :deal with insert select
         InsertStatement result = (InsertStatement) visit(ctx.insertValuesClause());
         result.setTable((SimpleTableSegment) visit(ctx.tableName()));
         result.setParameterCount(getCurrentParameterIndex());
@@ -230,7 +230,7 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     
     @Override
     public ASTNode visitSelect(final SelectContext ctx) {
-        // TODO : Unsupported for withClause.
+        // TODO :Unsupported for withClause.
         SelectStatement result = (SelectStatement) visit(ctx.unionClause());
         result.setParameterCount(getCurrentParameterIndex());
         return result;
@@ -238,7 +238,7 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     
     @Override
     public ASTNode visitUnionClause(final UnionClauseContext ctx) {
-        // TODO : Unsupported for union SQL.
+        // TODO :Unsupported for union SQL.
         return visit(ctx.selectClause(0));
     }
     
@@ -429,7 +429,7 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
-        // TODO: Bad processing for join table
+        // TODO :Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         ASTNode tableFactor = visit(ctx.tableFactor());
         if (tableFactor instanceof SubqueryTableSegment) {

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/PostgreSQLDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-postgresql/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/PostgreSQLDMLVisitor.java
@@ -393,6 +393,7 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     public ASTNode visitTableReference(final TableReferenceContext ctx) {
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         if (null != ctx.tableFactor()) {
+            // TODO :Ignore subquery table segment
             ASTNode tableFactor = visit(ctx.tableFactor());
             if (tableFactor instanceof SimpleTableSegment) {
                 result.getValue().add((SimpleTableSegment) tableFactor);
@@ -428,13 +429,17 @@ public final class PostgreSQLDMLVisitor extends PostgreSQLVisitor implements DML
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
+        // TODO: Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
-        SimpleTableSegment tableSegment = (SimpleTableSegment) visit(ctx.tableFactor());
-        result.getValue().add(tableSegment);
+        ASTNode tableFactor = visit(ctx.tableFactor());
+        if (tableFactor instanceof SubqueryTableSegment) {
+            return result;
+        }
+        result.getValue().add((SimpleTableSegment) tableFactor);
         if (null != ctx.joinSpecification()) {
             Collection<SimpleTableSegment> tableSegments = new LinkedList<>();
             for (SimpleTableSegment each : ((CollectionValue<SimpleTableSegment>) visit(ctx.joinSpecification())).getValue()) {
-                if (isTable(each, Collections.singleton(tableSegment))) {
+                if (isTable(each, Collections.singleton((SimpleTableSegment) tableFactor))) {
                     tableSegments.add(each);
                 }
             }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-relation/src/main/java/org/apache/shardingsphere/sql/parser/relation/segment/select/projection/engine/ProjectionsContextEngine.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-relation/src/main/java/org/apache/shardingsphere/sql/parser/relation/segment/select/projection/engine/ProjectionsContextEngine.java
@@ -65,7 +65,7 @@ public final class ProjectionsContextEngine {
      */
     public ProjectionsContext createProjectionsContext(final String sql, final SelectStatement selectStatement, final GroupByContext groupByContext, final OrderByContext orderByContext) {
         ProjectionsSegment projectionsSegment = selectStatement.getProjections();
-        Collection<Projection> projections = getProjections(sql, selectStatement.getTables(), projectionsSegment);
+        Collection<Projection> projections = getProjections(sql, selectStatement.getSimpleTableSegments(), projectionsSegment);
         ProjectionsContext result = new ProjectionsContext(projectionsSegment.getStartIndex(), projectionsSegment.getStopIndex(), projectionsSegment.isDistinctRow(), projections);
         result.getProjections().addAll(getDerivedGroupByColumns(projections, groupByContext, selectStatement));
         result.getProjections().addAll(getDerivedOrderByColumns(projections, orderByContext, selectStatement));
@@ -189,7 +189,7 @@ public final class ProjectionsContextEngine {
     }
     
     private SimpleTableSegment find(final String tableNameOrAlias, final SelectStatement selectStatement) {
-        for (SimpleTableSegment each : selectStatement.getTables()) {
+        for (SimpleTableSegment each : selectStatement.getSimpleTableSegments()) {
             if (tableNameOrAlias.equalsIgnoreCase(each.getTableName().getIdentifier().getValue()) || tableNameOrAlias.equals(each.getAlias().orElse(null))) {
                 return each;
             }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-relation/src/main/java/org/apache/shardingsphere/sql/parser/relation/statement/dml/SelectStatementContext.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-relation/src/main/java/org/apache/shardingsphere/sql/parser/relation/statement/dml/SelectStatementContext.java
@@ -83,7 +83,7 @@ public final class SelectStatementContext extends CommonSQLStatementContext<Sele
     public SelectStatementContext(final SelectStatement sqlStatement, final GroupByContext groupByContext,
                                   final OrderByContext orderByContext, final ProjectionsContext projectionsContext, final PaginationContext paginationContext) {
         super(sqlStatement);
-        tablesContext = new TablesContext(sqlStatement.getTables());
+        tablesContext = new TablesContext(sqlStatement.getSimpleTableSegments());
         this.groupByContext = groupByContext;
         this.orderByContext = orderByContext;
         this.projectionsContext = projectionsContext;
@@ -93,7 +93,7 @@ public final class SelectStatementContext extends CommonSQLStatementContext<Sele
     
     public SelectStatementContext(final RelationMetas relationMetas, final String sql, final List<Object> parameters, final SelectStatement sqlStatement) {
         super(sqlStatement);
-        tablesContext = new TablesContext(sqlStatement.getTables());
+        tablesContext = new TablesContext(sqlStatement.getSimpleTableSegments());
         groupByContext = new GroupByContextEngine().createGroupByContext(sqlStatement);
         orderByContext = new OrderByContextEngine().createOrderBy(sqlStatement, groupByContext);
         projectionsContext = new ProjectionsContextEngine(relationMetas).createProjectionsContext(sql, sqlStatement, groupByContext, orderByContext);
@@ -187,7 +187,7 @@ public final class SelectStatementContext extends CommonSQLStatementContext<Sele
     
     @Override
     public Collection<SimpleTableSegment> getAllTables() {
-        Collection<SimpleTableSegment> result = new LinkedList<>(getSqlStatement().getTables());
+        Collection<SimpleTableSegment> result = new LinkedList<>(getSqlStatement().getSimpleTableSegments());
         if (getSqlStatement().getWhere().isPresent()) {
             result.addAll(getAllTablesFromWhere(getSqlStatement().getWhere().get()));
         }
@@ -205,7 +205,7 @@ public final class SelectStatementContext extends CommonSQLStatementContext<Sele
         Collection<SimpleTableSegment> result = new LinkedList<>();
         for (AndPredicate each : where.getAndPredicates()) {
             for (PredicateSegment predicate : each.getPredicates()) {
-                result.addAll(new PredicateExtractor(getSqlStatement().getTables(), predicate).extractTables());
+                result.addAll(new PredicateExtractor(getSqlStatement().getSimpleTableSegments(), predicate).extractTables());
             }
         }
         return result;
@@ -222,7 +222,7 @@ public final class SelectStatementContext extends CommonSQLStatementContext<Sele
     
     private Optional<SimpleTableSegment> getTableSegment(final ProjectionSegment each) {
         Optional<OwnerSegment> owner = getTableOwner(each);
-        if (owner.isPresent() && isTable(owner.get(), getSqlStatement().getTables())) {
+        if (owner.isPresent() && isTable(owner.get(), getSqlStatement().getSimpleTableSegments())) {
             return Optional .of(new SimpleTableSegment(owner.get().getStartIndex(), owner.get().getStopIndex(), owner.get().getIdentifier()));
         }
         return Optional.empty();
@@ -243,7 +243,7 @@ public final class SelectStatementContext extends CommonSQLStatementContext<Sele
         for (OrderByItemSegment each : orderByItems) {
             if (each instanceof ColumnOrderByItemSegment) {
                 Optional<OwnerSegment> owner = ((ColumnOrderByItemSegment) each).getColumn().getOwner();
-                if (owner.isPresent() && isTable(owner.get(), getSqlStatement().getTables())) {
+                if (owner.isPresent() && isTable(owner.get(), getSqlStatement().getSimpleTableSegments())) {
                     Preconditions.checkState(((ColumnOrderByItemSegment) each).getColumn().getOwner().isPresent());
                     OwnerSegment segment = ((ColumnOrderByItemSegment) each).getColumn().getOwner().get();
                     result.add(new SimpleTableSegment(segment.getStartIndex(), segment.getStopIndex(), segment.getIdentifier()));

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-sql92/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQL92DMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-sql92/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQL92DMLVisitor.java
@@ -277,7 +277,7 @@ public final class SQL92DMLVisitor extends SQL92Visitor implements DMLVisitor {
     
     @Override
     public ASTNode visitProjection(final ProjectionContext ctx) {
-        // FIXME: The stop index of project is the stop index of projection, instead of alias.
+        // FIXME :The stop index of project is the stop index of projection, instead of alias.
         if (null != ctx.qualifiedShorthand()) {
             QualifiedShorthandContext shorthand = ctx.qualifiedShorthand();
             ShorthandProjectionSegment result = new ShorthandProjectionSegment(shorthand.getStart().getStartIndex(), shorthand.getStop().getStopIndex(), shorthand.getText());

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-sql92/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQL92DMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-sql92/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQL92DMLVisitor.java
@@ -196,7 +196,7 @@ public final class SQL92DMLVisitor extends SQL92Visitor implements DMLVisitor {
     
     @Override
     public ASTNode visitSelect(final SelectContext ctx) {
-        // TODO : Unsupported for withClause.
+        // TODO :Unsupported for withClause.
         SelectStatement result = (SelectStatement) visit(ctx.unionClause());
         result.setParameterCount(getCurrentParameterIndex());
         return result;
@@ -204,7 +204,7 @@ public final class SQL92DMLVisitor extends SQL92Visitor implements DMLVisitor {
     
     @Override
     public ASTNode visitUnionClause(final UnionClauseContext ctx) {
-        // TODO : Unsupported for union SQL.
+        // TODO :Unsupported for union SQL.
         return visit(ctx.selectClause(0));
     }
     
@@ -392,7 +392,7 @@ public final class SQL92DMLVisitor extends SQL92Visitor implements DMLVisitor {
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
-        // TODO: Bad processing for join table
+        // TODO :Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         ASTNode tableFactor = visit(ctx.tableFactor());
         if (tableFactor instanceof SubqueryTableSegment) {

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-sql92/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQL92DMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-sql92/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQL92DMLVisitor.java
@@ -356,6 +356,7 @@ public final class SQL92DMLVisitor extends SQL92Visitor implements DMLVisitor {
     public ASTNode visitTableReference(final TableReferenceContext ctx) {
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         if (null != ctx.tableFactor()) {
+            // TODO :Ignore subquery table segment
             ASTNode tableFactor = visit(ctx.tableFactor());
             if (tableFactor instanceof SimpleTableSegment) {
                 result.getValue().add((SimpleTableSegment) tableFactor);
@@ -391,13 +392,17 @@ public final class SQL92DMLVisitor extends SQL92Visitor implements DMLVisitor {
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
+        // TODO: Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
-        SimpleTableSegment tableSegment = (SimpleTableSegment) visit(ctx.tableFactor());
-        result.getValue().add(tableSegment);
+        ASTNode tableFactor = visit(ctx.tableFactor());
+        if (tableFactor instanceof SubqueryTableSegment) {
+            return result;
+        }
+        result.getValue().add((SimpleTableSegment) tableFactor);
         if (null != ctx.joinSpecification()) {
             Collection<SimpleTableSegment> tableSegments = new LinkedList<>();
             for (SimpleTableSegment each : ((CollectionValue<SimpleTableSegment>) visit(ctx.joinSpecification())).getValue()) {
-                if (isTable(each, Collections.singleton(tableSegment))) {
+                if (isTable(each, Collections.singleton((SimpleTableSegment) tableFactor))) {
                     tableSegments.add(each);
                 }
             }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-sqlserver/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQLServerDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-sqlserver/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQLServerDMLVisitor.java
@@ -379,6 +379,7 @@ public final class SQLServerDMLVisitor extends SQLServerVisitor implements DMLVi
     public ASTNode visitTableReference(final TableReferenceContext ctx) {
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         if (null != ctx.tableFactor()) {
+            // TODO :Ignore subquery table segment
             ASTNode tableFactor = visit(ctx.tableFactor());
             if (tableFactor instanceof SimpleTableSegment) {
                 result.getValue().add((SimpleTableSegment) tableFactor);
@@ -414,13 +415,17 @@ public final class SQLServerDMLVisitor extends SQLServerVisitor implements DMLVi
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
+        // TODO: Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
-        SimpleTableSegment tableSegment = (SimpleTableSegment) visit(ctx.tableFactor());
-        result.getValue().add(tableSegment);
+        ASTNode tableFactor = visit(ctx.tableFactor());
+        if (tableFactor instanceof SubqueryTableSegment) {
+            return result;
+        }
+        result.getValue().add((SimpleTableSegment) tableFactor);
         if (null != ctx.joinSpecification()) {
             Collection<SimpleTableSegment> tableSegments = new LinkedList<>();
             for (SimpleTableSegment each : ((CollectionValue<SimpleTableSegment>) visit(ctx.joinSpecification())).getValue()) {
-                if (isTable(each, Collections.singleton(tableSegment))) {
+                if (isTable(each, Collections.singleton((SimpleTableSegment) tableFactor))) {
                     tableSegments.add(each);
                 }
             }

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-sqlserver/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQLServerDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-sqlserver/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQLServerDMLVisitor.java
@@ -219,7 +219,7 @@ public final class SQLServerDMLVisitor extends SQLServerVisitor implements DMLVi
     
     @Override
     public ASTNode visitSelect(final SelectContext ctx) {
-        // TODO : Unsupported for withClause.
+        // TODO :Unsupported for withClause.
         SelectStatement result = (SelectStatement) visit(ctx.unionClause());
         result.setParameterCount(getCurrentParameterIndex());
         return result;
@@ -227,7 +227,7 @@ public final class SQLServerDMLVisitor extends SQLServerVisitor implements DMLVi
     
     @Override
     public ASTNode visitUnionClause(final UnionClauseContext ctx) {
-        // TODO : Unsupported for union SQL.
+        // TODO :Unsupported for union SQL.
         return visit(ctx.selectClause(0));
     }
     
@@ -415,7 +415,7 @@ public final class SQLServerDMLVisitor extends SQLServerVisitor implements DMLVi
     @SuppressWarnings("unchecked")
     @Override
     public ASTNode visitJoinedTable(final JoinedTableContext ctx) {
-        // TODO: Bad processing for join table
+        // TODO :Bad processing for join table
         CollectionValue<SimpleTableSegment> result = new CollectionValue<>();
         ASTNode tableFactor = visit(ctx.tableFactor());
         if (tableFactor instanceof SubqueryTableSegment) {

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-sqlserver/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQLServerDMLVisitor.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-sqlserver/src/main/java/org/apache/shardingsphere/sql/parser/visitor/impl/SQLServerDMLVisitor.java
@@ -300,7 +300,7 @@ public final class SQLServerDMLVisitor extends SQLServerVisitor implements DMLVi
     
     @Override
     public ASTNode visitProjection(final ProjectionContext ctx) {
-        // FIXME: The stop index of project is the stop index of projection, instead of alias.
+        // FIXME :The stop index of project is the stop index of projection, instead of alias.
         if (null != ctx.qualifiedShorthand()) {
             QualifiedShorthandContext shorthand = ctx.qualifiedShorthand();
             ShorthandProjectionSegment result = new ShorthandProjectionSegment(shorthand.getStart().getStartIndex(), shorthand.getStop().getStopIndex(), shorthand.getText());

--- a/shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dml/impl/SelectStatementAssert.java
+++ b/shardingsphere-sql-parser/shardingsphere-sql-parser-test/src/test/java/org/apache/shardingsphere/sql/parser/integrate/asserts/statement/dml/impl/SelectStatementAssert.java
@@ -63,7 +63,7 @@ public final class SelectStatementAssert {
     }
     
     private static void assertTable(final SQLCaseAssertContext assertContext, final SelectStatement actual, final SelectStatementTestCase expected) {
-        TableAssert.assertIs(assertContext, actual.getTables(), expected.getTables());
+        TableAssert.assertIs(assertContext, actual.getSimpleTableSegments(), expected.getTables());
     }
     
     private static void assertWhereClause(final SQLCaseAssertContext assertContext, final SelectStatement actual, final SelectStatementTestCase expected) {

